### PR TITLE
content/lxd: add new page for third-party integrations

### PR DESCRIPTION
**Pull request by toby63**
Closes: #503 

All logos are from official kits of the projects, with the exception of Ansible which is from the Wikipedia.

Signed-off-by: Tobias Gerold <tobias@g3ro.eu>

### CODE REVIEW

**Comment by toby63** [comment]
@stgraber Bad news, LXD seems to be deprecated (and replaced by LXC) on OpenNebula too: https://docs.opennebula.io/6.0/intro_release_notes/release_notes/compatibility.html#lxd-compatibility

I guess we need to check each tool again and then you need to decide how to proceed.

**Comment by toby63** [comment]
So here is an overview of what I could find for now:

| Software: | Supported: | Details: | Links: |
| --- | --- | --- | --- |
| Ansible | yes | recent documentation | https://docs.ansible.com/ansible/2.9/modules/lxd_container_module.html |
| Terraform | yes | Last update 8 months ago, so relatively recent.  | https://registry.terraform.io/providers/terraform-lxd/lxd/latest/docs |
| Puppet | yes (partly) | Supports only LXD 3.x and 2.x; last update june 2020 | https://github.com/ovh/lxd-puppet-module |
| Juju | yes  | recent documentation |  https://juju.is/docs/olm/lxd |
| OpenNebula | no (LXC is now supported instead) |   | https://docs.opennebula.io/6.0/open_cluster_deployment/lxc_node/overview.html <br> https://docs.opennebula.io/6.0/intro_release_notes/release_notes/compatibility.html#lxd |

**Comment by stgraber** [comment]
For puppet, it may be worth mentioning Puppet Bolt (https://puppet.com/docs/bolt/latest/bolt_transports_reference.html#lxd)

**Comment by stgraber** [comment]
I wasn't aware of OpenNebula moving away from LXD to using their own implementation on top of LXC, that disqualifies them for this list and also suggests we should be dropping the OpenNebula page from the website.

**Comment by stgraber** [comment]
So in short, I think we should:
 - Drop puppet from the list (out of tree plugin)
 - Add Puppet Bolt to the list (has mainline support)
 - Drop OpenNebula from the list
 - Drop the OpenNebula pages from the website

**Comment by stgraber** [comment]
Ah and we should probably add Packer to the list

**Comment by toby63** [comment]
Ok.

Regarding Packer: It looks like Packer is only for creating images for multiple platforms.
The startpage mentions terraform though:  `Use Terraform to launch completely provisioned and configured machine instances with Packer images in seconds.`
So is the usecase here only the image creation or more?

Regarding naming: I couldn't seem to find a good name for all this, so I named it "external integrations" for now, but I guess a better title might be...well...better.
I thought about "Hypervisor", but that is maybe not correct.

**Comment by toby63** [comment]
Additional question:
Can or should I add logos (whenever possible, because of licenses etc.)?

**Comment by stgraber** [comment]
That's correct, Packer is primarily used as a target-agnostic image builder. This makes it easy for someone to build images for AWS, GCP, ... as well as LXD.

I think using logos would be fine and may make the page look better.

Can't come up with something better than `External integrations` right now but maybe @tomponline or @ru-fu have some ideas :)

**Comment by ru-fu** [comment]
Tricky one ... :thinking: 

I'm not a fan of "External integrations", since my initial response to that is that LXD integrates external tools, and not that there are external tools that integrate LXD (which is what we mean here, correct?).

One alternative would be to name it "LXD integrations", but that still leaves some ambiguity. How about "External extensions" or "LXD wrappers"?

**Comment by tomponline** [comment]
@ru-fu @stgraber how about "ecosystem"?

**Comment by stgraber** [comment]
Having the page named Ecosystem would let us have integrations in both directions be listed.

So both external tools that know how to talk to LXD as we're covering here, but could also cover LXD integrating with things like MAAS, Candid and Canonical RBAC.

I guess the question is whether it sounds too generic to the point that visitors won't think to go see it?

**Comment by stgraber** [comment]
Another option could be "Third party integrations"? That would it more obvious that the focus would be on other projects driving LXD.

**Comment by ru-fu** [comment]
Ecosystem fits, but it's very generic. If we plan to do both directions, I like "Third-party integrations" (with hyphen).

**Comment by stgraber** [comment]
@tomp opinion on "Third-party integrations"?

**Comment by tomponline** [comment]
> @tomponline opinion on "Third-party integrations"?

Works for me :)

**Comment by stgraber** [comment]
Alright, let's go with that then!

**Comment by toby63** [comment]
Ok, then I will name it "third-party integrations".

**Comment by toby63** [comment]
The page is almost done, so comments are welcome.

Preview of the page (print of webpage as .pdf):
[Linux Containers - LXD - Getting started - Third-party integrations.pdf](https://github.com/lxc/linuxcontainers.org/files/7148527/Linux.Containers.-.LXD.-.Getting.started.-.Third-party.integrations.pdf)

One more thing:
- The documentation for Puppet Bolt is kind of sparse, so I found this repo that is mentioned in [the forum](https://discuss.linuxcontainers.org/t/puppet-bolt-lxd-transport-wip/9689/): https://github.com/dontlaugh/bolt-lxd-lab
   Should I link to that as well?

**Comment by stgraber** [comment]
Looking through this, one quick comment is that it's `a LXD server`, not `an LXD server` as LXD is pronounced lex-dee and not l-x-d.

**Comment by stgraber** [comment]
It feels a it repetitive to mention getting started, installation, ... for every one of them.

# Ansible
Ansible is an open-source software provisioning, conﬁguration management, and
application-deployment tool.

The main integrations between Ansible and LXD are:

 - A [connection plugin](https://docs.ansible.com/ansible/latest/collections/community/general/lxd_connection.html#ansible-collections-community-general-lxd-connection) allowing Ansible to manage a LXD instance just as any other Linux systems
 - An [inventory plugin](https://docs.ansible.com/ansible/latest/collections/community/general/lxd_inventory.html) to detect LXD instances
 - A [plugin](https://docs.ansible.com/ansible/latest/collections/community/general/lxd_container_module.html) to create and manage LXD containers
 - A [plugin](https://docs.ansible.com/ansible/latest/collections/community/general/lxd_profile_module.html) to manage LXD profiles

# Juju
Juju is an open-source software for application management and automation.

Juju can be used to deploy a variety of workloads across many different clouds and virtualization providers.
It supports both deploying workloads against a LXD server or cluster and using LXD on the machines it's deploying to separate otherwise colocated services.

Take a look at the Step-by-Step Guide in the Juju Documentation.

# Terraform
Terraform is an open-source infrastructure-as-code software tool for conﬁguration and service management.

The LXD integration allows for Terraform to deploy instances on LXD servers with support for local and remote deployments.

# Puppet Bolt
Bolt is an open source orchestration tool that automates the manual work it takes to maintain your infrastructure.

The LXD transport allows for interacting with LXD instances.

For more information, see the Puppet Bolt Documentation on LXD.

# Packer
Packer is an open source tool for creating identical machine images for multiple
platforms.

With Packer's LXD builder, it's possible to re-use your existing cloud image building pipeline and with the LXD builder, build a LXD container image.

**Comment by toby63** [comment]
> Looking through this, one quick comment is that it's a LXD server, not an LXD server as LXD is pronounced lex-dee and not l-x-d.

Good to know.

> It feels a it repetitive to mention getting started, installation, ... for every one of them.

Thats true :smile:, thanks for the input.
But still I would like to mention getting started etc., maybe I can put that into one section, like `General Notes` or `Set up LXD` and then I just link to that from every software section, e.g.:
`To manage LXD in Ansible, you need an LXD Server, see "Set up LXD"(link to specific header).`

**Comment by stgraber** [comment]
Yeah, probably worth using the Puppet Bolt logo.

I don't think we need to care too much about copyright/trademark notices on those, though if any of the trademark holders are upset about our use of their logo, we'll definitely be happy to remove their logo or mention of their product from the page :)

**Comment by toby63** [comment]
@stgraber Ready for another review.

File: content/lxd/introduction.md [hunk 1]
```diff
@@ -20,6 +20,12 @@ You can find a series of howtos and tutorials on YouTube:
 
 <iframe width="560" height="315" src="https://www.youtube.com/embed/videoseries?list=PLddduKsl-KEhleT9VTR4hbtlNdtMr6cFd" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
 
+## Third-party integrations
```
Comment (stgraber) [inline content/lxd/introduction.md:23]: I think I'd move this section to be after the Features section.

**Comment by stgraber** [comment]
Just a tiny detail for the change to the instruction page, otherwise looks good to me.

File: content/lxd/introduction.md (hunk 1 above)
Comment (toby63) [reply content/lxd/introduction.md]: I moved it after `Availability`, so that the regular installation methods get displayed first.

**Comment by stgraber** [comment]
Thanks!

**Pull request merged**